Build a regression model.

Before building the model I want to better classify the POI 

In [47]:
%store -r bike_station_information_df
%store -r yelp_result_df
%store -r fourSquare_result_df
%store -r final_df

final_df = final_df

  
def categorize(category):
  # Restaurants
  if any(keyword in category for keyword in ['Bakery', 'Restaurant', 'French', 'Japanese', 'Pizza', 'Indian', 'Pan Asian', 'Moroccan', 'Thai', 'Bistros', 'Pizzeria', 'Burger Joint', 'Steakhouses', 'Italian', 'Seafood', 'Mexican', 'Chinese', 'Fast Food']):
    return 'Restaurant'
  
  # Bars and Lounges
  elif any(keyword in category for keyword in ['Bar', 'Pub', 'Lounge']):
    return 'Bar/Lounge'
  
  # Health Centers
  elif any(keyword in category for keyword in ['Hospital', 'Health', 'Drugstore']):
    return 'Health Center'
  
  # Retails
  elif any(keyword in category for keyword in ['Computers and Electronics Retail', 'Store', 'Shop', 'Retail', 'Bookstore', 'Botanical Gardens']):
    return 'Retail'
  
  # Education
  elif any(keyword in category for keyword in ['School', 'Education', 'Library']):
    return 'Education'
  
  # Entertainment
  elif any(keyword in category for keyword in ['Music Venue', 'Theater', 'Cinema', 'Arts and Entertainment', 'Paintball Field']):
    return 'Entertainment'
  
  # Home Services
  elif any(keyword in category for keyword in ['Carpenter', 'Home Improvement Service', 'Electrician', 'Plumber', 'Interior Designer', 'Roof Deck', 'Heating, Ventilating and Air Conditioning Contractor']):
    return 'Home Services'
  
  # Landmarks and Historical Buildings
  elif any(keyword in category for keyword in ['Monument', 'Scenic Lookout', 'Park', 'Landmarks & Historical Buildings', 'Lakes']):
    return 'Landmarks & Historical Buildings'
  
  # Business and Professional Services
  elif any(keyword in category for keyword in ['Management Consultant', 'Business and Professional Services', 'Employment Agency', 'Organization', 'Professional Cleaning Service', 'Accounting and Bookkeeping Service']):
    return 'Business Services'
  
  # Other Services
  elif any(keyword in category for keyword in ['Computer Repair Service', 'Audiovisual Service', 'Post Office', 'Utility Company']):
    return 'Other Services'
  
  elif 'N/A' in category:
    return 'N/A'
  
  # Miscellaneous
  else:
    return 'Miscellaneous'


final_df['GroupedCategory'] = final_df['Category'].apply(categorize)

# Checking for all unique values in the 'grouped Category' column
unique_new_categories = final_df['GroupedCategory'].unique()
print(unique_new_categories)


final_df[final_df['Key'] == 0]

['Education' 'Retail' 'Restaurant' 'N/A' 'Health Center' 'Bar/Lounge'
 'Other Services' 'Miscellaneous' 'Landmarks & Historical Buildings'
 'Business Services' 'Home Services' 'Entertainment']


City   Latitude  Longitude  Number of Bikes  Key  \
0    Cergy  49.034625   2.067596               12    0   
1    Cergy  49.034625   2.067596               12    0   
2    Cergy  49.034625   2.067596               12    0   
3    Cergy  49.034625   2.067596               12    0   
4    Cergy  49.034625   2.067596               12    0   
5    Cergy  49.034625   2.067596               12    0   
6    Cergy  49.034625   2.067596               12    0   
7    Cergy  49.034625   2.067596               12    0   
8    Cergy  49.034625   2.067596               12    0   
9    Cergy  49.034625   2.067596               12    0   
440  Cergy  49.034625   2.067596               12    0   
441  Cergy  49.034625   2.067596               12    0   
442  Cergy  49.034625   2.067596               12    0   
443  Cergy  49.034625   2.067596               12    0   
444  Cergy  49.034625   2.067596               12    0   
445  Cergy  49.034625   2.067596               12    0   
446  Cergy  49.034625   2.067596               12    0   
447  Cergy  49.034625   2.067596               12    0   
448  Cergy  49.034625   2.067596               12    0   
449  Cergy  49.034625   2.067596               12    0   

                                     Name                   Category  \
0          Ecole de biologie industrielle                  Education   
1                            Esprit Clean  Fashion Accessories Store   
2                    Le Millésime du Port          French Restaurant   
3                     La Taverne des Rois                 Restaurant   
4                     Au Fourmont Village                     Bakery   
5                      Moulin Vert Lava'9                        N/A   
6                        SAN MARINA Cergy                        N/A   
7    École Maternelle Publique le Village          Elementary School   
8                   France Depannage Auto        Miscellaneous Store   
9               Pharmacie des Trois Gares                  Drugstore   
440                       La Rotisserie Ô                     French   
441                          Hippopotamus                Steakhouses   
442                          La Cavallina                    Italian   
443                   Les Relais d'Alsace                     French   
444                           O'Sullivans                       Pubs   
445                       Ici et Ailleurs                     French   
446                        Le Port d'Asie                       Thai   
447               Au festival de la Moule                    Seafood   
448                       Relais d'alsace                     French   
449                             Le Latino                    Mexican   

        Distance  Rating GroupedCategory  
0     259.000000     NaN       Education  
1     283.000000     NaN          Retail  
2     634.000000     NaN      Restaurant  
3     648.000000     NaN      Restaurant  
4     359.000000     NaN      Restaurant  
5     391.000000     NaN             N/A  
6     394.000000     NaN             N/A  
7     397.000000     NaN       Education  
8     425.000000     NaN          Retail  
9     448.000000     NaN   Health Center  
440   500.015071     4.5      Restaurant  
441   592.851567     2.5      Restaurant  
442   496.288387     5.0      Restaurant  
443   638.741161     4.0      Restaurant  
444   823.376072     3.0      Bar/Lounge  
445  1121.515453     4.0      Restaurant  
446   495.942094     4.0      Restaurant  
447   518.464532     3.0      Restaurant  
448   625.861077     3.0      Restaurant  
449   489.924766     1.0      Restaurant

In [46]:
import pandas as pd

# Creating a new dataframe using information from the combined dataframe (final_df)
grouped_df = pd.pivot_table(final_df, 
                            index=['Key', 'Number of Bikes', 'Latitude', 'Longitude'],
                            columns='GroupedCategory', 
                            values='Category', 
                            aggfunc='count', 
                            fill_value=0).reset_index()
grouped_df.columns.name = None

grouped_df

Key  Number of Bikes   Latitude  Longitude  Bar/Lounge  Business Services  \
0     0               12  49.034625   2.067596           1                  0   
1     1                0  49.036320   2.079568           6                  0   
2     2                4  49.017697   2.099417           0                  0   
3     3                8  49.044092   2.036546           0                  1   
4     4               14  49.030665   2.083067           3                  2   
5     5               12  49.035511   2.061921           0                  0   
6     6                4  49.052785   2.017956           0                  0   
7     7               10  49.038058   2.070534           4                  1   
8     8                6  49.044395   2.090632           2                  0   
9     9                2  49.038874   2.076405           5                  1   
10   10                6  49.043583   2.084808           2                  0   
11   11                5  49.018057   2.091026           0                  0   
12   12                4  49.042999   2.029687           0                  1   
13   13                8  49.043284   2.068934           3                  0   
14   14               13  49.031214   2.061506           4                  0   
15   15                7  49.048246   2.011238           0                  1   
16   16                4  49.030106   2.021573           0                  0   
17   17                4  49.011216   2.098463           0                  1   
18   18                7  49.024210   2.103268           1                  0   
19   19                7  49.032107   2.074345           5                  0   
20   20                7  49.046868   2.013620           0                  1   
21   21                8  49.014534   2.078945           0                  0   
22   22                4  49.049401   2.017075           0                  1   
23   23                9  49.049300   2.035223           1                  0   
24   24                6  49.046588   2.045120           0                  0   
25   25               15  49.032141   2.087802           3                  2   
26   26                2  49.023163   2.029486           0                  0   
27   27                7  49.042318   2.019580           1                  0   
28   28                4  49.048957   2.026947           2                  0   
29   29                6  49.041233   2.075956           4                  0   
30   30                2  49.051599   2.036105           1                  0   
31   31                7  49.023837   2.022595           0                  1   
32   32                6  49.043224   2.060405           1                  0   
33   33                8  49.047781   2.103038           1                  0   
34   34                5  49.051104   2.069419           1                  0   
35   35                2  49.023065   2.110460           1                  1   
36   36                6  49.050988   2.057324           0                  0   
37   37               10  49.032230   2.101656           1                  0   
38   38                3  49.032934   2.020312           0                  0   
39   39                3  49.045223   2.002507           0                  1   
40   40                3  49.027239   2.077178           0                  1   
41   41                4  49.046249   2.105710           2                  1   
42   42                9  49.047619   2.096823           1                  1   
43   43                4  49.044831   2.053462           0                  0   

    Education  Entertainment  Health Center  Home Services  \
0           2              0              1              0   
1           0              0              0              0   
2           3              0              1              0   
3           1              0              0              1   
4           3              0              0     

Provide model output and an interpretation of the results. 

In [48]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

X = grouped_df[['Latitude', 'Longitude', 'Bar/Lounge', 'Business Services', 'Education', 'Entertainment',
        'Health Center', 'Home Services', 'Landmarks & Historical Buildings', 'Miscellaneous', 'N/A',
        'Other Services', 'Restaurant', 'Retail']]
y = grouped_df['Number of Bikes']

X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:        Number of Bikes   R-squared:                       0.419
Model:                            OLS   Adj. R-squared:                  0.139
Method:                 Least Squares   F-statistic:                     1.494
Date:                Tue, 12 Dec 2023   Prob (F-statistic):              0.175
Time:                        23:10:12   Log-Likelihood:                -103.92
No. Observations:                  44   AIC:                             237.8
Df Residuals:                      29   BIC:                             264.6
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
const   

The R-squared value of 0.419 indicates that the model explains 41.9% of the variability in the Number of Bikes. However, the lower Adjusted R-squared value (0.139) suggests that not all included variables are contributing significantly to the model.

Among the independent varaibles, only Longitude, Landmarks and Historical buildings and restaurant have a P value less than 0.05 which would suggest that these variables may have a statistically significant relationship with the Number of Bikes. Variables with p-values greater than 0.05 may not be contributing significantly and could be considered for removal from the model. 

The latitude and longitude seem to have the highest coefficient value which would suggest they have an higher impact on the number of bikes 

# Stretch

How can you turn the regression model into a classification model?